src/lib.rs

```diff
@@ -705,15 +743,55 @@ impl<EF: ExtraTokenFields, TT: TokenType, TE: ErrorResponseType> Client<EF, TT,
         (self.authorize_url_impl("code", Some(&state), None), state)
     }
 
+    ///
+    /// Produces the full authorization URL used by the
+    /// [Proof Key for Code Exchange](https://tools.ietf.org/html/rfc7636) extension to the
+    /// [Authorization Code Grant](https://tools.ietf.org/html/rfc6749#section-4.1) flow.
+    ///
+    /// # Arguments
+    ///
+    /// * `state_fn` - A function that returns an opaque value used by the client to maintain state
+    ///   between the request and callback. The authorization server includes this value when
+    ///   redirecting the user-agent back to the client.
+    /// * `verifier_fn` - A function that returns a PKCE code challenge.
+    ///
+    /// # Security Warning
+    ///
+    /// Callers should use a fresh, unpredictable `state` for each authorization request and verify
+    /// that this value matches the `state` parameter passed by the authorization server to the
+    /// redirect URI. Doing so mitigates
+    /// [Cross-Site Request Forgery](https://tools.ietf.org/html/rfc6749#section-10.12)
+    /// attacks.
+    ///
+    /// Callers should use a fresh, unpredictable `code_verifier` for each authorization request.
+    ///
+    pub fn authorize_url_pkce<F, G>(&self, state_fn: F, verifier_fn: G)
+        -> (Url, CsrfToken, CodeVerifier)
+    where
+        F: FnOnce() -> CsrfToken,
+        G: FnOnce() -> CodeVerifier
+    {
+        let state = state_fn();
+        let verifier = verifier_fn();
+        let digest = Sha256::digest(verifier.secret().as_bytes());
+        let code_challenge = base64::encode_config(&digest, base64::URL_SAFE_NO_PAD);
+        let params = vec![
+            ("code_challenge_method", "S256"),
```

Comment:
It might be nice to support other methods to allow for future extensibility, but it's probably not worth complicating the API. I'm guessing S256 will remain the dominant method. In any case, callers can use `authorize_url_extension` if they want to use a different method. I agree with not supporting `plain` at all here.